synapse/handlers/auth.py

```diff
@@ -494,6 +523,17 @@ def _get_session_info(self, session_id: Optional[str]) -> dict:
 
         return self.sessions[session_id]
 
+    def _invalidate_session(self, session_id) -> None:
+        """Invalidate session information for session ID"""
+        session = self.sessions.get(session_id, None)
+        if session and "ui_auth" in session:
+            # Set the items in the ui_auth session to sentinel values that can
```

Comment:
can we not just delete the session from `self.sessions`?

Reply:
This gets called when the UI auth session is completed, but before it has been "used" (the client has to re-request whatever resources initially kicked off the UI-auth in the first place, I believe). Fully removing the session breaks this.

The test in [tests/rest/client/v2_alpha/test_auth.py](https://github.com/matrix-org/synapse/blob/4ce50519cdfe482ef5833488295f0235ad9fe0a9/tests/rest/client/v2_alpha/test_auth.py) fail if you remove the session which clued me into this.

Reply:
The docstring should probably note something about when this is expected to be called...

Reply:
Actually, I think that test is failing regardless of this change. I'll re-investigate!

Reply:
OK. I retraced some of the logic here and this might be the wrong place to invalidate sessions. This logic path seems to be followed:

1. When the final flow of auth is completed.
2. When the request is re-done to actually perform the action.

The action fails since it is invalidated right after auth completes.

Reply:
This was a legit bug. Nice find! I believe I have a fix, but need to make a couple of other changes. Thanks!